lib/connection/connection.js

```diff
@@ -528,97 +528,142 @@ function merge(options1, options2) {
-      return undefined;
+function prepareConnectionOptions(self, _options) {
+  let options;
+  if (self.ssl) {
+    options = {
+      socket: self.connection,
+      rejectUnauthorized: self.rejectUnauthorized
     };
-  } else if (typeof self.checkServerIdentity === 'function') {
-    sslOptions.checkServerIdentity = self.checkServerIdentity;
-  }
 
-  // Set default sni servername to be the same as host
-  if (sslOptions.servername == null) {
-    sslOptions.servername = self.host;
-  }
+    // Merge in options
+    merge(options, self.options);
+    merge(options, _options);
+
+    // Set options for ssl
+    if (self.ca) options.ca = self.ca;
+    if (self.crl) options.crl = self.crl;
+    if (self.cert) options.cert = self.cert;
+    if (self.key) options.key = self.key;
+    if (self.passphrase) options.passphrase = self.passphrase;
+
+    // Override checkServerIdentity behavior
+    if (self.checkServerIdentity === false) {
+      // Skip the identiy check by retuning undefined as per node documents
+      // https://nodejs.org/api/tls.html#tls_tls_connect_options_callback
+      options.checkServerIdentity = function() {
+        return undefined;
+      };
+    } else if (typeof self.checkServerIdentity === 'function') {
+      options.checkServerIdentity = self.checkServerIdentity;
+    }
 
-  // Attempt SSL connection
-  const connection = tls.connect(self.port, self.host, sslOptions, function() {
-    // Error on auth or skip
-    if (connection.authorizationError && self.rejectUnauthorized) {
-      return self.emit('error', connection.authorizationError, self, { ssl: true });
+    // Set default sni servername to be the same as host
+    if (options.servername == null) {
+      options.servername = self.host;
     }
 
+    options = Object.assign({}, options, { host: self.host, port: self.port });
+
+    return options;
+  } else {
+    if (self.domainSocket) {
+      options = { path: self.host };
+    } else {
+      options = { port: self.port, host: self.host };
+    }
+    return options;
+  }
+}
+
+function makeConnection(self, options, callback) {
+  const module = options.ssl ? tls : net;
+
+  const connection = module.connect(options, function() {
+    if (self.ssl) {
+      // Error on auth or skip
+      if (connection.authorizationError && self.rejectUnauthorized) {
+        return self.emit('error', connection.authorizationError, self, { ssl: true });
+      }
+    }
     // Set socket timeout instead of connection timeout
+
     connection.setTimeout(self.socketTimeout);
-    // We are done emit connect
-    self.emit('connect', self);
+    return callback(null, connection);
   });
 
   // Set the options for the connection
   connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
   connection.setTimeout(self.connectionTimeout);
   connection.setNoDelay(self.noDelay);
 
-  return connection;
+  // Add handlers for events
+  connection.once('error', err =>
```

Comment:
for all of these I think its enough to just call `callback(err, null)` . Either we have an error or we have a connection, we don't really care where the error came from at this point, or what type it is.